src/app/jscaip/MGPNode.ts

```diff
@@ -1,268 +1,185 @@
+        let label: string = `#${this.gameState.turn}: ${this.id}`;
+        if (labelFn) {
+            label += ` - ${labelFn(this)}`;
         }
-    }
-    public myToString(): string {
-        let genealogy: string = '';
-        // eslint-disable-next-line @typescript-eslint/no-this-alias
-        let node: MGPNode<R, M, S, L, B> = this;
-        if (node.mother.isAbsent()) {
-            const turn: number = node.gameState.turn;
-            return 'InitialNode: ' + turn;
+        console.log(`    node_${id} [label="${label}", style=filled, fillcolor="${color}"];`);
+
+        let nextId: number = id+1;
+        if (max === undefined || level < max) {
+            for (const child of this.children.listValues()) {
+                console.log(`    node_${id} -> node_${nextId} [label="${child.move.get()}"];`);
+                nextId = child.printDot(rules, labelFn, max, level+1, nextId);
+            }
         }
-        while (node.mother.isPresent()) {
-            const move: string = node.move.isAbsent() ? ' ' : ' > ' + node.move.get().toString() + '> ';
-            const turn: number = node.gameState.turn;
-            genealogy = move + turn + ' ' + genealogy;
-            node = node.mother.get();
+        if (level === 0) {
+            console.log('}');
         }
-        return 'Node: ' + genealogy;
+        return nextId;
     }
-    public hasMoves(): boolean {
-        return this.childs.isPresent();
+    /**
+     * Get a value from the cache, or MGPOptional if it does not exist in the cache.
+     */
+    public getCache<T>(key: string): MGPOptional<T> {
+        return this.cache.get(key) as MGPOptional<T>;
     }
-    // debug
-    public countDescendants(): number {
-        if (this.childs.isAbsent()) {
-            return 0;
-        }
-        let nbDescendants: number = this.childs.get().length;
-        for (const son of this.childs.get()) {
-            nbDescendants += son.countDescendants();
+    /**
+     * Set or replace a value from the cache.
+     */
+    public setCache(key: NonNullable<string>, value: NonNullable<unknown>): void {
+        if (this.cache.containsKey(key)) {
+            this.cache.replace(key, value);
+        } else {
+            this.cache.set(key, value);
         }
-        return nbDescendants;
     }
 }
 
-export abstract class AbstractNode extends MGPNode<Rules<Move, GameState, unknown>, Move, GameState, unknown> {
+export class AbstractNode extends GameNode<Move, GameState> {}
+
+// TODO FOR REVIEW: à partir d'ici, le reste appartiendrait à AI.ts
+
+/**
+ * A move generator should have a method that generates move from a node.
+ * It may generate all possible moves, but may also just filter out some uninteresting moves.
+ */
+export abstract class MoveGenerator<M extends Move, S extends GameState> {
+    /**
+     * Gives the list of all the possible moves.
+     * Has to be implemented for each rule so that the AI can choose among theses moves.
```

Comment:
how early does this fail otherwise, not sure we already throw UT error when no Minimax are set for a game

Reply:
pas sûr de comprendre ce que tu veux dire !

Reply:
pas avant (pour ça qu'on a eu récemment un jeu (teeko?) sans IA) mais maintenant oui